src/network/networkd-sysctl.c

```diff
@@ -40,10 +42,82 @@ static void manager_set_ip_forwarding(Manager *manager, int family) {
                                   enable_disable(t), af_to_ipv4_ipv6(family));
 }
 
+static void sysctl_overwrite_check(sd_event_source *source, const struct inotify_event *event, Hashmap *sysctls) {
+        _cleanup_free_ char *buf = NULL, *path = NULL;
+        const char *dir_path, *value;
+        size_t len;
+        int r;
+
+        assert(event);
+
+        r = sd_event_source_get_inotify_path(source, &dir_path);
+        if (r < 0)
+                return;
+
+        path = path_join(dir_path, event->name);
+        if (!path)
+                return;
+
+        value = hashmap_get(sysctls, path);
+
+        /* We don't handle this sysctl */
+        if (!value)
+                return;
+
+        if (read_full_file(path, &buf, &len) < 0)
+                return;
+
+        delete_trailing_chars(buf, "\n");
+
+        if (!streq(value, buf))
```

Comment:
instead of the `delete_trailing_chars()` thing just use `streq_skip_trailing_chars(value, buf, /* ok= */ WHITESPACE)`.

That way you compare the two strings, but ignore trailing whitespace on *both* strings

Reply:
But in this case "buf" will still contain a trailing \n, which breaks the warning:

```
Jun 23 01:17:13 localhost systemd-networkd[399]: sysctl /proc/sys/net/ipv4/conf/enp0s1/promote_secondaries was changed from '1' to '2
Jun 23 01:17:13 localhost systemd-networkd[399]: '
```

Reply:
i see, but I think i'd still use streq_skip_trailing_chars() here, and then just call delete_trailing_chars(buf, …) when invoking log_warning, on the fly.

i.e:
```c
if (!streq_skip_trailing_chars(value, buf, /* ok= */ WHITESPACE))
        log_warning("sysctl %s was changed from '%s' to '%s'", path, value, delete_trailing_chars(buf, "\n"));
```

Hmm, also, I think we probably should say "changed externally" rather than just "changed", to emphasize that this wasn't us.

in fact, I think we should actually make this a recognizable, structued log messages, i.e. one with log_struct() that uses MESSAGE_ID= and comes with an entry in catalog/systemd.catalog.in that explains the situation, i.e. mentions that the conflict of ownership.

Reply:
The current approach looks OK to me, if we store values with trailing newlines in the shadow hashmap something has gone seriously wrong so we shouldn't have to handle that scenario here.